pkg/runner/test_runner.go

```diff
@@ -158,234 +159,301 @@ func newTargetWriter(log *logrus.Entry, timeouts TestRunnerTimeouts) *targetWrit
 
 	// `targets` is used to buffer targets coming from the previous routing blocks,
 	// queueing them for injection into the TestStep. The list is accessed
 	// synchronously by a single goroutine.
 	targets := list.New()
 
-	// `ingressTarget` and `egressTarget` are used to keep track of ingress and
-	// egress times and perform sanity checks on the input/output of the TestStep
+	// `ingressTarget` is used to keep track of ingress times of a target into a test step
 	ingressTarget := make(map[*target.Target]time.Time)
-	egressTarget := make(map[*target.Target]time.Time)
 
-	var (
-		err                           error
-		stepInClosed, routeInProgress bool
-		injectionWg                   sync.WaitGroup
-		numIngress                    uint64
-	)
+	// Channel that the injection goroutine uses to communicate back to `routeIn` the results
+	// of asynchronous injection
+	injectResultCh := make(chan injectionResult)
 
-	for {
-		log.Debugf("listening")
+	// injectionChannels are used to inject targets into test step and return results to `routeIn`
+	injectionChannels := injectionCh{stepIn: r.routingChannels.stepIn, resultCh: injectResultCh}
 
+	log.Debugf("initializing routeIn for %s", stepLabel)
+	targetWriter := newTargetWriter(log, r.timeouts)
+
+	for {
 		select {
-		case <-terminateRoute:
+		case <-terminate:
 			err = fmt.Errorf("termination requested for routing into %s", stepLabel)
-			break
 		case injectionResult := <-injectResultCh:
 			log.Debugf("received injection result for %v", injectionResult.target)
 			routeInProgress = false
 			if injectionResult.err != nil {
 				err = fmt.Errorf("routing failed while injecting target %+v into %s", injectionResult.target, stepLabel)
 				targetInErrEv := testevent.Data{EventName: target.EventTargetInErr, Target: injectionResult.target}
-				if err := ev.Emit(targetInErrEv); err != nil {
+				if err := r.ev.Emit(targetInErrEv); err != nil {
 					log.Warningf("could not emit %v event for target: %+v", targetInErrEv, *injectionResult.target)
 				}
 			} else {
 				targetInEv := testevent.Data{EventName: target.EventTargetIn, Target: injectionResult.target}
-				if err := ev.Emit(targetInEv); err != nil {
+				if err := r.ev.Emit(targetInEv); err != nil {
 					log.Warningf("could not emit %v event for Target: %+v", targetInEv, *injectionResult.target)
 				}
 			}
-		case t, chanIsOpen := <-tRouteIn:
+		case t, chanIsOpen := <-r.routingChannels.routeIn:
 			if !chanIsOpen {
 				log.Debugf("routing input channel closed")
-				tRouteIn = nil
-				if position == 0 {
-					log.Debugf("setting in targets: %d", len(ingressTarget))
-					atomic.StoreUint64(&p.numIngress, numIngress)
-				}
+				r.routingChannels.routeIn = nil
 			} else {
 				log.Debugf("received target %v in input", t)
-				numIngress++
 				targets.PushFront(t)
 			}
-		case t, chanIsOpen := <-tStepOut:
+		}
+
+		if err != nil {
+			break
+		}
+
+		if routeInProgress {
+			continue
+		}
```

Comment:
When we receive something from `routeIn`, we might not want to continue, and instead proceed with injection of the target that we just received, if there is no other target pending. We still need a way to tell whether there is already a goroutine pending for the injection. Then, we need to check this status in two situations:

1) We receive another target
2) We finish injecting another target

In case of 1), we need to check if another injection is pending. In case 2) we know that `routeInProgress` needs to be reset, there is nothing being injected. 
I don't think I have found a clean way to get rid of that boolean flag.

The defer approach looks fine, I can amend in that direction, but I don't think I have found a clean way to get rid of that boolean flag.

Reply:
It seems to me a too difficult logic for a too simple task.

If I understand correctly, the task is: we want to forward targets from `r.routingChannels.routeIn` to `r.routingChannels.stepIn` without queueing (to `r.routingChannels.routeIn`). With timeouts and stuff, but this is currently non relevant :)

Few questions about the current code:
1. Why do we need targets list at all? Isn't it just a duplication of channels logic (of `routeIn` in this case)? (which could queue things)?
2. Couldn't we just do something like this? -- https://play.golang.org/p/Kvki6Z2RNHm

Reply:
Please let me know if comments to the code into the link above will help. I'll elaborate :)

Reply:
Sorry for the long delay, I had taken notes that I needed to reply to this but got around doing so only now :( 


> If I understand correctly, the task is: we want to forward targets from r.routingChannels.routeIn to r.routingChannels.stepIn without queueing (to r.routingChannels.routeIn).

Yes, this is the intention.

1. We need a target list to be used as buffer to accept as soon as possible incoming targets. We could use a channel, but we would need that channel to automatically resize. We don't know beforehand the number of targets that we might want to enqueue. Proposal in 2. assumes a buffer size of 10. Which size should we set here?
3. In this unit tests for this control path have already been added in this PR. In particular, `TestRouteInRoutesAllTargets` tests that all targets incoming from `routeIn`, go out from `stepIn` (where the unit test is consuming targets). However, here your solution would pass this unit test. What I think is important is that anything writing to `routeIn` never blocks (because on the other side somebody is always reading on the channel). This was one of the assumptions when designing the pipeline, and the reason is that I wanted for targets to be taken into charge from plugins as soon as possible. So if a plugin returns a target, the routing block needs to accept it without making the test step block, under no circumstance. This is made possible by having that list, which acts as a buffer. With a channel with fixed size, we might end up in a situation were we are blocked on channel write. And if  this is the contract with the plugins, I think it might make things harder to debug when things timeout or we seem to occur into a deadlock.

Reply:
> We don't know beforehand the number of targets that we might want to enqueue. Proposal in 2. assumes a buffer size of 10. Which size should we set here?

1. We do know how may target we have acquired.
2. Even if in the proposal you remove `, 10` from `make(chan` it still will work correctly :)

> With a channel with fixed size, we might end up in a situation were we are blocked on channel write

This only means that nobody wants to receive anything from that channel, yet (just like if you remove `, 10` in the proposal). It is nothing really solved by creating a buffer: you just move the queue from the initial point to another one.

---

But I think I understood what you mean:
A pipeline consists of test steps, and we want to:
* Make logic of test steps as simple as possible.
* Finish a step as soon as possible.
This we want to read all targets from the previous step as soon as possible and push it to next step on its convenience. So we have to create this temporary buffer or predict a queue size to create channels of the requires size. My thoughts here:
1. I do not understand why it is important to finish a step as soon as possible. For neater logging? I'm not sure if that justifies this logic difficulty in a so responsible part of the code.
2. More importantly, we know the amount of targets returned by the target manager and we could've use this.

---

But I think it does not worth so thorough discussion and let's just leave it as is :)

Reply:
Quick comment:

> More importantly, we know the amount of targets returned by the target manager and we could've use this.

There are two complications:
* Target might fail while flowing through the pipeline, so we know the maximum number
* Pipelines can be attached back to back. A second pipeline does not know how many target will flow through it until the first pipeline has completed, so in this case we really need to wait a signal from outside telling us that no more targets will come through. 


I could use the upper bound in both cases, probably it's a better tradeoff. 

Let me propose this: I'll write a solution with using the maximum target number that could ever flow through the pipeline. And we discuss in a different PR if that is a good enough tradeoff.